File: Week 02/id_682/590.n-ary-tree-postorder-traversal.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""
class Solution(object):
    def postorder(self, root):
        
        """
        :type root: Node
        :rtype: List[int]
        """
        
        if root is None:
            return []
        
        res = []
        stack = [root]
        
        while stack:
            temp = stack.pop()
            
            res.append(temp.val)
            
            if temp.children:
                for val in temp.children:
                    stack.append(val)
        
        return res[::-1]

File: Week 02/id_662/590.n-ary-tree-postorder-traversal.java
/**
 * 给定一个 N 叉树，返回其节点值的后序遍历。
 * <p>
 * 例如，给定一个 3叉树 :
 * <p>
 * <p>
 * <p>
 * <p>
 * <p>
 * <p>
 * <p>
 * 返回其后序遍历: [5,6,3,2,4,1].
 *
 */

class Solution {
    public List<Integer> postorder(Node root) {
        List<Integer> list = new ArrayList<Integer>();
        if(root==null){
            return list;
        }
        return postorderHelper(root,list);
    }

    public List<Integer> postorderHelper(Node root,List<Integer> list) {
        List<Node> children = root.children;
        for(Node node:children){
            postorderHelper(node,list);
        }
        list.add(root.val);
        return list;
    }
}